concurrency/LocalMUCRoom.java

```diff
@@ -0,0 +1,3450 @@
+        JID affiliationJID = affiliation.getBareJID();
+        switch(affiliation.getAffiliation()) {
+            case owner:
+                removeMember(affiliationJID);
+                removeAdmin(affiliationJID);
+                removeOutcast(affiliationJID);
+                owners.add(affiliationJID);
+                addRegisteredUser(affiliationJID.toString()); // *** ONE LINE AMAZON
+                break;
+            case admin:
+                removeMember(affiliationJID);
+                removeOwner(affiliationJID);
+                removeOutcast(affiliationJID);
+                admins.add(affiliationJID);
+                addRegisteredUser(affiliationJID.toString()); // *** ONE LINE AMAZON
+                break;
+            case outcast:
+                removeMember(affiliationJID);
+                removeAdmin(affiliationJID);
+                removeOwner(affiliationJID);
+                outcasts.add(affiliationJID);
+                break;
+            case none:
+            default:
+                removeMember(affiliationJID);
+                removeAdmin(affiliationJID);
+                removeOwner(affiliationJID);
+                removeOutcast(affiliationJID);
+                break;
+        }
+    }
+
+    @Override
+    public void nicknameChanged(MUCRole occupantRole, Presence newPresence, String oldNick, String newNick) {
+        // Ask other cluster nodes to update the nickname of the occupant
+        ChangeNickname request = new ChangeNickname(this, oldNick,  newNick, newPresence.createCopy());
+        CacheFactory.doClusterTask(request);
+
+        // Update the nickname of the occupant
+        request = new ChangeNickname(this, oldNick,  newNick, newPresence.createCopy());
+        request.setOriginator(true);
+        request.run();
+
+        // Broadcast new presence of occupant
+        broadcastPresence(occupantRole.getPresence().createCopy(), false);
+    }
+
+    public void nicknameChanged(ChangeNickname changeNickname) {
+        List<MUCRole> occupants = occupantsByNickname.get(changeNickname.getOldNick().toLowerCase());
+        if (occupants != null && occupants.size() > 0) {
+            for (MUCRole occupant : occupants) {
+                // Update the role with the new info
+                occupant.setPresence(changeNickname.getPresence());
+                occupant.changeNickname(changeNickname.getNewNick());
+            }
+            if (changeNickname.isOriginator()) {
+                // Fire event that user changed his nickname
+                MUCEventDispatcher.nicknameChanged(getRole().getRoleAddress(), occupants.get(0).getUserAddress(),
+                        changeNickname.getOldNick(), changeNickname.getNewNick());
+            }
+            // Associate the existing MUCRole with the new nickname
+            occupantsByNickname.put(changeNickname.getNewNick().toLowerCase(), occupants);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()` and `put()` may not be thread-safe at lines: **2487, 2488, 2489, and 2500**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).